File: ruby/etl/etl.rb
module ETL
  def self.transform(value)
    pairs = value.flat_map do |score, letters|
      pairs_of(score, letters)
    end
    Hash[*pairs.flatten]
  end

  private

  def self.pairs_of(score, letters)
    letters.map { |l| [l.downcase, score] }
  end
end

File: haskell/etl/src/ETL.hs
module ETL (transform) where

import Data.Map (Map, fromList, assocs)
import Data.Char (toLower)

transform :: Map a String -> Map Char a
transform legacyData = fromList scorePairs
  where 
    scorePairs = concatMap transformWords $ assocs legacyData

transformWords :: (a, String) -> [(Char, a)]
transformWords (score, letters) = map (\c -> (toLower c, score)) letters
